NASA-PDS/validate
# Fix bug with printStackTraceToFile() function and make stack printing…

… optional with new command line parameter --enable-stack-printing

1. Add test resource :src/test/resources/github379
2. Add test resource :src/test/resources/github380 (Directory is required due to quirkiness with behave)
3. Add logic to enable stack printing :src/main/java/gov/nasa/pds/tools/util/FileService.java
4. Add house keeping of stackPrintingFlag variable :src/main/java/gov/nasa/pds/tools/util/FlagsUtil.java
5. Fix bug with printStackTraceToFile(0 function :src/main/java/gov/nasa/pds/tools/validate/rule/pds4/TableDataContentValidationRule.java
6. Add parsing ENABLE_STACK_PRINTING parameter :src/main/java/gov/nasa/pds/validate/ValidateLauncher.java
7. Add ENABLE_STACK_PRINTING parameter :src/main/java/gov/nasa/pds/validate/commandline/options/Flag.java
8. Add ENABLE_STACK_PRINTING option :src/main/java/gov/nasa/pds/validate/commandline/options/FlagOptions.java
9. Add tests for github379 and github380 :src/test/resources/features/validate.feature

Refs:

https://github.com/nasa-pds/validate/issues/379 FileService:printStackTraceToFile:ERROR when validating a product with overlapping fields
https://github.com/nasa-pds/validate/issues/380 stack trace being created during successful validate execution

Closes:

https://github.com/nasa-pds/validate/issues/379 FileService:printStackTraceToFile:ERROR when validating a product with overlapping fields  
https://github.com/nasa-pds/validate/issues/380 stack trace being created during successful validate execution

Tested in DEV:

For issue #379: There should be no crashing.  The report file will properly report there was an issue with 
```
"error.table.field_value_overlap"
              "severity": "ERROR",
              "type": "error.table.field_value_overlap",

% validate -R pds4.label --skip-context-reference-check -r report_github379_label_invalid.json   -s json -t  src/test/resources/github379/mix_cal_hk_fpac_report_20181204.xml  
```
For issue #380: There should be no crashing.  The report file will properly report there was an issue with
```
"error.table.field_value_overlap"
              "severity": "ERROR",
              "type": "error.table.field_value_overlap",

# This test re-use resources from github379
% validate -R pds4.label --skip-context-validation --skip-context-reference-check -r report_github380_label_invalid_without_stack.json   -s json -t  src/test/resources/github379/mix_cal_hk_fpac_report_20181204.xml  >& t2
```

If the --debug-mode is used, the stack will be written to a stack file with the name validate_stack_traces.log
Make sure the file is delete before running next test
```
% rm validate_stack_traces.log
% validate -R pds4.label --debug-mode --skip-context-validation --skip-context-reference-check -r report_github380_label_invalid_with_stack.json   -s json -t  src/test/resources/github380/mix_cal_hk_fpac_report_20181204.xml  >& t3
```
